### PR TITLE
NIFI-3204 fix handling deleting a path with a wildcard when the processor is in…

### PR DESCRIPTION
**Pull request by francoisprunier**
…ssor is invoqued via an incoming flowfile

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
Yes
- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/DeleteHDFS.java [hunk 1]
```diff
@@ -141,20 +141,33 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
+
                     if (!context.hasIncomingConnection()) {
                         flowFile = session.create();
+                        session.transfer(session.putAllAttributes(flowFile, attributes), REL_FAILURE);
                     }
+
+                }
+            }
+            if (context.hasIncomingConnection()) {
+                // TODO we only put the last path deleted, change the semantic of the processor?
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/DeleteHDFS.java]: We should remove this 'TODO' and instead document the behavior you're describing on the capability decription/and/or the WritesAttribute section.  If it is important for a user to see a listing of deleted items they should not use the wildcard here but rather use ListHDFS to create an explicit set of files and this processor should be honoring precisely the items specified.

File: nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/DeleteHDFS.java [hunk 2]
```diff
@@ -141,20 +141,33 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
...
+                if (!foundMissingFile) {
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/DeleteHDFS.java]: Looks like 'foundMissingFile' is set to false above and then there is this if/else.  Perhaps this is incomplete?  Goes with the 'todo'?

**Comment by joewitt** [comment]
The more I look at this processor the more I think the mistake was in allowing it to have wildcards and to run without incoming flow files.  Had it been restricted to simply deleting the content of a path specified by a given flowfile it would be much more clear.

**Comment by joewitt** [comment]
@francoisprunier @pvillard31 Hello.  I've attached a patch file to the JIRA https://issues.apache.org/jira/browse/NIFI-3204.  The documentation of the processor did not declare its behavior one way or another though tests suggested it created flowfiles for things it deleted.  Instead, I have made the behavior and logic more explicit and simplified and updated to the documentation and tests to align to that.  For more detailed knowledge of what was deleted/etc.. users should take advantage of this processor in combination with ListHDFS.

The patch file I provided applies overtop this PR as of commit aeab1feeb3c5f47858793e00425b26cc9f2d5b0b.  You can apply the patch on top of your PR and advise if this takes care of the problem and you agree with the fix.

Thanks

**Comment by francoisprunier** [comment]
Hi @joewitt,

I've successfully tested your patch. The thing I'm concerned about is the change of behaviour when no incoming flow is connected to the processor, which breaks the contract existing users of the processor may rely on.

Before:
- incoming flow: creates one flow file per deleted file: BROKEN when more than one file deleted
- no incoming flow: creates one flow file per deleted file per run: WORKS

After:
- incoming flow: "creates" only one flow file per incoming flowfile: WORKS when more than one file deleted
- no incoming flow: creates only one flow file per run: WORKS but behaviour is changed

On the one hand I like your approach better, but on the other hand breaking users' flows is not great.

**Comment by joewitt** [comment]
@francoisprunier sorry for the delay in responding.  Will try to come back to this soon!

**Comment by francoisprunier** [comment]
@joewitt, I've applied your patch and pushed it, we should be good to go.

**Pull request closed without merging**